### DISCUSSION TITLE
Energy Meter write Modbus Data

### DISCUSSION THREAD

**Discussion by timaseth** [arendst/Tasmota#14810] [Q&A] [answered]
I have a modbus energy meter for which I am using the Smart Meter Interface. Works very well.
Is it possible to send data to a few Holding Register (which accept both read and write) to reset energy count and such?
Not many, just a few.
Thanks in advance!

**Comment by gemu2015** [comment, marked as answer]
the very last tasmota dev version supports this.
#define USE_SML_SCRIPT_CMD

define a string variable in >D with the MODBUS cmd you want to send
take "raw" mode to send any format (crc is calculated inside function) e.g.
str="r0442"

in any other section you may send the string to the meter by
sml(1 3 str) 
1 being the meter number, 3 is the selector for that function

**Reply by Erickrk** [reply]
Hey there, I've had some doubts regarding your answer: when using the "raw" mode I should write the full frame by myself, right? e.g. str="r0106272e001e637f"
When you're talking about the "selector for that function" would it be function 03, reading in Modbus, or a specific Smart Meter function?

Do you have a working script using the write function?

**Reply by Erickrk** [reply]
I've found [this](https://tasmota.github.io/docs/Scripting-Language/#special-variables) link on the documentation that explains the "sml(1 3 str)" part, my second question. But I still lack a functional code.

**Reply by gemu2015** [reply]
this is a working example of a growatt solar inverter how to set the export limit

\>D 22
cstr=""
gl=0
tmp=0
\>B
=>sensor53 r
\>S
if chg[gl]>0 {
	; change limit
	tmp=int(gl/42)
	cstr="r0106000300"+hn(tmp)
	sml(1 3 cstr)
}
\>M 1
+1,18,m,0,9600,GRW,19,5,01040026,01040028,01040005,01040009,01030003
1,010404UUuu@i0:10,Netzspannung,V,mainsv,1
1,010404xxxxUUuu@i0:10,Einspeisestrom,A,mainsc,1
1,010404UUuuUUuu@i1:10,Einspeiseleistung,W,mainsw,1
1,010404UUuuUUuu@i2:10,string 1 unten,W,s1w,1
1,010404UUuuUUuu@i3:10,string 2 oben,W,s2w,1
1,010304UUuu@i4:1,limit,%,limit,0
\#
\>W
`<hr>`
nm(1000 3600 10 gl "Growatt limit (W) " 80 0)

**Reply by Erickrk** [reply]
Thanks a lot, @gemu2015 ! I understood it all and my script is working great now.

**Reply by jensrohloff** [reply]
@gemu2015 did you publish some more details to your growatt setup somewhere else? Are you able to use the export limit via modbus to control Nulleinspeisung?

**Reply by gemu2015** [reply]
newer growatt inverters have 2 modbus interfaces, one slave to read the operating values and one master to control the export limit. you may find a special manual on the growatt support pages how to wire that.

in principle you need a modbus 3 phase meter (they recommend SDM630) in serial with your grid meter that is connected via wire to your growatt inverter. in the growatt setup you then may set the export limit to 70 percent of module peak power or even zero for "nulleinspeisung".

you may also manage that via the other growatt interface and the above script. also in this case you need the actual exported power to set the growatt limit.  you then need to adjust the actual export limit so that nothing is exported. this control loop has to be fast enough to hold the actual export low enough.

i doubt however that the german TÜV will accept that home made solution while the inbuild growatt solution is certified.

**Reply by jensrohloff** [reply]
Thanks a lot @gemu2015 the info is actually very helpful to complete my solution.

**Comment by SpareEnergie** [comment]
Thank you, @gemu2015, your code showed me a way. However, the growatt hold register is written via a second interface that is decoupled via diodes and writes to the TX of the Modbus module. Here is a snippet.
>B
so(13 15 9600)
=>sensor53 r
>M 1......
>S
array[1]=01
array[2]=06
array[3]=00
array[4]=03
array[5]=00
array[6]=limit
swa(array 8 (0))
...
The required limit is calculated from the current power from the E-meter, which is made available via another ESP using global variables.
This means that an acceptable zero control now works.

Questions:
1. Can the hold register be written to indefinitely?
If eeproms are installed, the limit is 100T.
2. If the limit is increased suddenly, the Growatt needs more than 2 minutes to reach 50% of its power, for example.
Which registers can be used to increase the speed at which the power increases?

**Comment by SpareEnergie** [comment]
Further research showed that the hold registers are stored in the EEPROMs. In my opinion, this means that power control via the "Active Power Rate" register 3 is no longer useful.

As no SDM630 v2 is available, I tried to emulate an SDM630 using the existing SDM72 via the registers requested by Growatt. This is very roughly achieved by changing the register 30013 Phase 1 active power.

A Modbus recording of an SDM630 v2 v3 during power control would therefore be desirable. Thank you

**Reply by gemu2015** [reply]
register 3 limit is not stored in  EEPROM, it is lost after restart and meant for dynamic power limit,
permanent limit ist done via these registers:
<img width="719" alt="Bildschirmfoto 2024-12-14 um 07 28 57" src="https://github.com/user-attachments/assets/781b97ef-cac8-40ec-b7cd-b722b943e12c" />

i can not confirm that limitation takes 2 minutes, it adapts immediately.
